src/main/java/com/iota/iri/storage/rocksDB/RocksDBPPPImpl.java

```diff
@@ -0,0 +1,607 @@
+package com.iota.iri.storage.rocksDB;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.iota.iri.controllers.TransactionViewModel;
+import com.iota.iri.model.Hash;
+import com.iota.iri.model.persistables.Transaction;
+import com.iota.iri.model.persistables.Tag;
+import com.iota.iri.model.persistables.Bundle;
+import com.iota.iri.model.persistables.Address;
+import com.iota.iri.model.persistables.Approvee;
+import com.iota.iri.model.persistables.Hashes;
+import com.iota.iri.storage.PermanentPersistenceProvider;
+import com.iota.iri.storage.Indexable;
+import com.iota.iri.storage.Persistable;
+import com.iota.iri.storage.PersistenceProvider;
+import com.iota.iri.utils.IotaIOUtils;
+import com.iota.iri.utils.Pair;
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.lang3.SystemUtils;
+
+import org.rocksdb.util.SizeUnit;
+import org.rocksdb.ColumnFamilyHandle;
+import org.rocksdb.RocksDB;
+import org.rocksdb.DBOptions;
+import org.rocksdb.BloomFilter;
+import org.rocksdb.WriteBatch;
+import org.rocksdb.WriteOptions;
+import org.rocksdb.RocksEnv;
+import org.rocksdb.BlockBasedTableConfig;
+import org.rocksdb.StringAppendOperator;
+import org.rocksdb.MergeOperator;
+import org.rocksdb.ColumnFamilyOptions;
+import org.rocksdb.ColumnFamilyDescriptor;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.nio.ByteBuffer;
+import java.nio.file.Paths;
+import java.util.Set;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+
+/***
+ * Implements the retrieval functions of a the normal persistence provider and the permanent persistence provider
+ */
+public class RocksDBPPPImpl implements PermanentPersistenceProvider, PersistenceProvider {
```

Comment:
Now having a deeper look here, I would like to offer a different design :-)

Currently, it is possible with the current design it is theoretically possible to implement a new `PersistanceProvider` that will replace RocksDb. You have actually attempted to do it yourself :-)

Here you tightly coupled RocksDb with the PermenantPersistanceProvider...
This means that if we ever want to replace RocksDB it will be much harder (there are actually thoughts about doing this now - #1661).

Also I have noticed lots of evil copy paste... Here we are believers in DRY (don't repeat yourself).

So maybe you can create `PermenantPersistanceProvider` that contains a `RocksDbPersistanceProvider`?
This way if we write a new PersistnaceProvider, switching will be a breeze?

It will also adhere better to the single responsibility principle that we have a class that only deals with wrapping the RocksDb api and a class that deals with the logic of your task :-)